python/tank/platform/validation.py

```diff
@@ -379,16 +389,31 @@ def validate(self, settings):
         # first sanity check that the schema is correct
         validate_schema(self._display_name, self._schema)
         
-        # Ensure that all required keys are in the settings and that the
-        # values are appropriate.
+        # Ensure that all keys are in the settings or have a default value in
+        # the manifest. Also make sure values are appropriate.
         for settings_key in self._schema:
             value_schema = self._schema.get(settings_key, {})
-            
-            # make sure the required key exists in the environment settings
-            if settings_key not in settings:
-                raise TankError("Missing required key '%s' in settings!" % settings_key)
-            
-            self.__validate_settings_value(settings_key, value_schema, settings[settings_key])
+
+            if settings_key in settings:
+                # value exists in the settings. use it.
+                settings_value = settings[settings_key]
+            else:
+                # No value in the settings. Look for a default value.
+                if value_schema.get("type") == "hook":
+                    # Hooks already have the concept of a default value. If this
+                    # setting is a hook, just use that special default hook
+                    # string and allow the validation code do its thing.
+                    settings_value = constants.TANK_BUNDLE_DEFAULT_HOOK_SETTING
```

Comment:
@manneohrstrom - This all makes sense. I'm still trying to wrangle backward compatibility. We need to still support `default` as a valid hook value in the environments, and we need to support `primary_publish` style hook defaults in the manifests...

With sparse configs, when the setting is asked for, it will fall back to the value in the manifest if not specified in the environment. For hooks, this means it would return say, `primary_publish` for an old-style hook whereas in the past it would have returned the magic `default` value added to the settings at install/update time. Now, when the hook is executed, the magic `default` path is resolved as the bundle's `hooks` directory, but `primary_publish` value from the manifest is resolved to the config's `hooks` directory. 

So there has to be something special that happens to old-style hook default values in order to maintain backward compatibility I think. So either return the magic `default` value or prepend `{self}`. Does that make sense? Maybe I'm overthinking things. 

I'll push my latest WIP shortly and maybe that'll help frame my thought process.